### PR TITLE
Testing mode

### PR DESCRIPTION
**Pull request by jbinz**

### CODE REVIEW

File: firmware/src/config/cats_config.cpp [hunk 1]
```diff
@@ -59,7 +60,8 @@ const cats_config_t DEFAULT_CONFIG = {
     .battery_type = LI_ION,
     .rec_speed_idx = 0,
     /* Assume that when the user starts the board for the first time the default config will be considered theirs. */
-    .is_set_by_user = true};
+    .is_set_by_user = true,
+    .enable_testing_mode = false};
```
Comment (stojadin2701) [inline firmware/src/config/cats_config.cpp:64]: move this above, is_set_by_user should be last

File: firmware/src/cli/settings.cpp [hunk 2]
```diff
@@ -135,7 +138,12 @@ const cli_value_t value_table[] = {
     {"battery_type", VAR_UINT8 | MODE_LOOKUP, {.lookup = {TABLE_BATTERY}}, offsetof(cats_config_t, battery_type)},
 
     {"rec_elements", VAR_UINT32, {.u32_max = UINT32_MAX}, offsetof(cats_config_t, rec_mask)},
-    {"rec_speed", VAR_UINT8 | MODE_LOOKUP, {.lookup = {TABLE_SPEEDS}}, offsetof(cats_config_t, rec_speed_idx)}};
+    {"rec_speed", VAR_UINT8 | MODE_LOOKUP, {.lookup = {TABLE_SPEEDS}}, offsetof(cats_config_t, rec_speed_idx)},
+    {"enable_test_mode",
```
Comment (stojadin2701) [inline firmware/src/cli/settings.cpp:142]: test_mode or testing_mode?

File: firmware/src/config/cats_config.hpp [hunk 3]
```diff
@@ -47,6 +47,9 @@ struct cats_config_t {
   battery_type_e battery_type;
   uint8_t rec_speed_idx;  // == inverse recording rate - 1
   bool is_set_by_user;
+
+  /* Testing Mode */
+  bool enable_testing_mode;
```
Comment (stojadin2701) [inline firmware/src/config/cats_config.hpp:52]: move above

File: firmware/src/tasks/task_telemetry.hpp [hunk 4]
```diff
@@ -40,23 +40,38 @@ class Telemetry final : public Task<Telemetry, 1024> {
     int16_t velocity : 10;
     uint16_t voltage : 8;
     uint16_t pyro_continuity : 2;
+    uint8_t testing_on : 1;
```
Comment (stojadin2701) [inline firmware/src/tasks/task_telemetry.hpp:43]: bool

File: firmware/src/tasks/task_telemetry.hpp [hunk 5]
```diff
@@ -40,23 +41,38 @@ class Telemetry final : public Task<Telemetry, 1024> {
...
+    uint32_t passcode;
+    uint8_t event;
+    uint8_t enable_testing_telemetry;
+    uint32_t dummy1;
+    uint32_t dummy2;
+  } __attribute__((packed));
+
+  uint32_t m_uplink_phrase_crc = 0;
+  bool m_testing_enabled;
+  bool m_enable_testing_telemetry = false;
```
Comment (l-jost) [inline firmware/src/tasks/task_telemetry.hpp:61]: Maybe rename to something like:
```suggestion
  bool m_testing_armed = false;
```

File: firmware/src/tasks/task_telemetry.hpp [hunk 6]
```diff
@@ -40,23 +41,38 @@ class Telemetry final : public Task<Telemetry, 1024> {
...
+  struct packed_rx_msg_t {
+    uint8_t header;
+    uint32_t passcode;
+    uint8_t event;
+    uint8_t enable_testing_telemetry;
+    uint32_t dummy1;
+    uint32_t dummy2;
+  } __attribute__((packed));
+
+  uint32_t m_uplink_phrase_crc = 0;
```
Comment (stojadin2701) [inline firmware/src/tasks/task_telemetry.hpp:59]: add comments to all of these

File: firmware/src/tasks/task_telemetry.hpp [hunk 7]
```diff
@@ -40,23 +41,38 @@ class Telemetry final : public Task<Telemetry, 1024> {
...
+  bool m_testing_enabled;
+  bool m_enable_testing_telemetry = false;
+  static constexpr uint32_t RX_PACKET_HEADER{0x72};
```
Comment (stojadin2701) [inline firmware/src/tasks/task_telemetry.hpp:62]: explain why 0x72 (if it's random say it's random)

File: firmware/src/tasks/task_telemetry.cpp [hunk 8]
```diff
@@ -87,7 +87,7 @@ void Telemetry::PackTxMessage(uint32_t ts, gnss_data_t* gnss, packed_tx_msg_t* t
 
   tx_payload->lat = static_cast<int32_t>(gnss->position.lat * 10000);
   tx_payload->lon = static_cast<int32_t>(gnss->position.lon * 10000);
-  tx_payload->testing_on = static_cast<uint8_t>(m_testing_enabled);
+  tx_payload->testing_on = static_cast<bool>(m_testing_enabled);
```
Comment (stojadin2701) [inline firmware/src/tasks/task_telemetry.cpp:90]: isn't this a bool already?

File: firmware/src/tasks/task_telemetry.hpp [hunk 9]
```diff
@@ -41,25 +41,26 @@ class Telemetry final : public Task<Telemetry, 1024> {
...
 
   struct packed_rx_msg_t {
-    uint8_t header;
-    uint32_t passcode;
-    uint8_t event;
-    uint8_t enable_testing_telemetry;
+    uint8_t header;    /* Header used to check if the packet is used for arming */
+    uint32_t passcode; /* Testing passcode, only if this matches with the configured passcode is the packet accepted. */
+    uint8_t event;     /* Event which needs to be triggered */
+    uint8_t enable_testing_telemetry; /* If this bit is set to one, the flight computer arms itself for testing; only
```
Comment (stojadin2701) [inline firmware/src/tasks/task_telemetry.hpp:54]: bool?

File: firmware/src/tasks/task_telemetry.hpp [hunk 10]
```diff
@@ -41,25 +41,26 @@ class Telemetry final : public Task<Telemetry, 1024> {
...
     uint8_t : 0;  // sent
     uint8_t d1;   // dummy
   } __attribute__((packed));
 
   struct packed_rx_msg_t {
-    uint8_t header;
-    uint32_t passcode;
-    uint8_t event;
-    uint8_t enable_testing_telemetry;
+    uint8_t header;    /* Header used to check if the packet is used for arming */
```
Comment (stojadin2701) [inline firmware/src/tasks/task_telemetry.hpp:51]: put comments above 
/* ... */
uint8_t header;
<new line>
/* ... */
uint32_t passcode;
...

File: firmware/src/tasks/task_telemetry.cpp [hunk 11]
```diff
@@ -58,6 +59,10 @@ void Telemetry::PackTxMessage(uint32_t ts, gnss_data_t* gnss, packed_tx_msg_t* t
     tx_payload->state = m_fsm_enum;
   }
 
+  if (m_testing_enabled) {
+    tx_payload->state = static_cast<uint8_t>(m_testing_armed);
```
Comment (stojadin2701) [inline firmware/src/tasks/task_telemetry.cpp:63]: what's going on here?

File: firmware/src/tasks/task_telemetry.hpp [hunk 12]
```diff
@@ -40,23 +41,42 @@ class Telemetry final : public Task<Telemetry, 1024> {
...
   } __attribute__((packed));
 
+  struct packed_rx_msg_t {
+    /* Header used to check if the packet is used for arming */
+    uint8_t header;
+    /* Testing passcode, only if this matches with the configured passcode is the packet accepted. */
+    uint32_t passcode;
+    /* Event which needs to be triggered */
+    uint8_t event;
+    /* If this bit is set to one, the flight computer arms itself for testing; only then can events be triggered */
```
Comment (stojadin2701) [inline firmware/src/tasks/task_telemetry.hpp:57]:
```suggestion
    /* If this value is set to true, the flight computer arms itself for testing; only then can events be triggered */
```

**Review by l-jost — APPROVED** [review]

**Pull request merged**